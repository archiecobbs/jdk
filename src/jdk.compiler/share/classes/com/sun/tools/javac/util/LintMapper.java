/*
 * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package com.sun.tools.javac.util;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Consumer;
import java.util.stream.Stream;

import javax.tools.DiagnosticListener;
import javax.tools.JavaFileObject;

import com.sun.tools.javac.code.Lint;
import com.sun.tools.javac.code.Symbol;
import com.sun.tools.javac.tree.JCTree;
import com.sun.tools.javac.tree.JCTree.*;
import com.sun.tools.javac.tree.TreeInfo;
import com.sun.tools.javac.tree.TreeScanner;
import com.sun.tools.javac.util.JCDiagnostic.DiagnosticPosition;

/**
 * Maps source code positions to the applicable {@link Lint} instance based on {@link -Xlint}
 * command line flags and {@code @SuppressWarnings} annotations on containing declarations.
 *
 * <p>
 * This mapping can't be cannot be calculated until after attribution. As each top-level
 * declaration (class, package, or module) is attributed, this singleton is notified by
 * Attr and the {@link Lint}s contained in that declaration are calculated.
 *
 * <p><b>This is NOT part of any supported API.
 * If you write code that depends on this, you do so at your own risk.
 * This code and its internal interfaces are subject to change or
 * deletion without notice.</b>
 */
public class LintMapper {

    // The key for the context singleton
    private static final Context.Key<LintMapper> CONTEXT_KEY = new Context.Key<>();

    // Per-source file lint information
    private final Map<JavaFileObject, FileInfo> fileInfoMap = new HashMap<>();

    // Compiler context
    private final Context context;

    // The root Lint instance, calculated on-demand to avoid init loops
    private Lint rootLint;

    /**
     * Obtain the {@link LintMapper} context singleton.
     */
    public static LintMapper instance(Context context) {
        LintMapper instance = context.get(CONTEXT_KEY);
        if (instance == null)
            instance = new LintMapper(context);
        return instance;
    }

    /**
     * Constructor.
     */
    @SuppressWarnings("this-escape")
    protected LintMapper(Context context) {
        context.put(CONTEXT_KEY, this);
        this.context = context;
    }

    private Lint rootLint() {
        if (rootLint == null)
            rootLint = Lint.instance(context);
        return rootLint;
    }

    /**
     * Determine if the given file is known to this instance.
     *
     * @param sourceFile source file
     * @return true if file is recognized
     */
    public boolean isKnown(JavaFileObject sourceFile) {
        return fileInfoMap.containsKey(sourceFile);
    }

    /**
     * Obtain the {@link Lint} configuration that applies at the given position, if known.
     *
     * @param sourceFile source file
     * @param pos source position
     * @return the applicable {@link Lint}, if known
     */
    public Optional<Lint> lintAt(JavaFileObject sourceFile, DiagnosticPosition pos) {

        // If the file is completely unknown, we don't know
        FileInfo fileInfo = fileInfoMap.get(sourceFile);
        if (fileInfo == null)
            return Optional.empty();

        // If the file hasn't been fully parsed yet, we don't know what Lint applies yet
        if (!fileInfo.parsed)
            return Optional.empty();

        // Find the top-level declaration that contains pos; if there is none, then the root lint applies
        Span declSpan = fileInfo.findDeclSpan(pos);
        if (declSpan == null)
            return Optional.of(rootLint());

        // Have we attributed this top-level declaration? If not, we don't know what Lint applies yet
        List<LintSpan> lintSpans = fileInfo.lintSpanMap.get(declSpan);
        if (lintSpans == null)
            return Optional.empty();

        // Find the narrowest containing LintSpan; if there is none, then the root lint applies
        return FileInfo.bestMatch(lintSpans, pos)
          .map(lintSpan -> lintSpan.lint)
          .or(() -> Optional.of(rootLint()));
    }

    /**
     * Calculate {@lint Lint} configurations for all positions within the given top-level declaration.
     *
     * @param sourceFile source file
     * @param tree top-level declaration (class, package, or module)
     */
    public void calculateLints(JavaFileObject sourceFile, JCTree tree) {

//System.err.println("LintMapper.calculateLints("+tree.getTag()+"):"
//+"\n  sourceFile="+sourceFile
//+"\n  tree="+tree.getTag()+": " + new Span(tree)
//);

        // Get the info for this file
        FileInfo fileInfo = fileInfoMap.get(sourceFile);

        // Sanity checks
        Assert.check(isTopLevelDecl(tree));
        Assert.check(fileInfo != null && fileInfo.parsed);
        Span declSpan = new Span(tree);
        Assert.check(fileInfo.lintSpanMap.containsKey(declSpan), "unknown declaration");
        Assert.check(fileInfo.lintSpanMap.get(declSpan) == null, "duplicate calculateLints()");

        // Build the list of lints for declarations within the top-level declaration
        fileInfo.lintSpanMap.put(declSpan, new LintCalculator().calculate(tree, rootLint()));
    }

    /**
     * Reset this instance (except for listeners).
     */
    public void clear() {
        fileInfoMap.clear();
    }

    /**
     * Invoked when file parsing starts to create an entry for the new file.
     */
    public void startParsingFile(JavaFileObject sourceFile) {

//System.err.println("LintMapper.startParsingFile(): " + sourceFile);

        //Assert.check(!fileInfoMap.containsKey(sourceFile), "duplicate addSourceFile()");
        fileInfoMap.put(sourceFile, new FileInfo());
    }

    /**
     * Invoked when file parsing completes to identify the top-level declarations.
     */
    public void finishParsingFile(JCCompilationUnit tree) {

        // Get info for this file
        FileInfo fileInfo = fileInfoMap.get(tree.sourcefile);
        Assert.check(fileInfo != null, () -> "unknown source " + tree.sourcefile);
        Assert.check(!fileInfo.parsed, () -> "source already parsed: " + tree.sourcefile);
        Assert.check(fileInfo.lintSpanMap.isEmpty(), () -> "duplicate invocation for " + tree.sourcefile);

        // Mark file as parsed
        fileInfo.parsed = true;

        // Create an entry in lintSpanMap for each top-level declaration, with a null value for now
        tree.defs.stream()
          .filter(this::isTopLevelDecl)
          .map(Span::new)
          .forEach(span -> fileInfo.lintSpanMap.put(span, null));

//System.err.println("LintMapper.finishParsingFile(): " + tree.sourcefile + " -> " + fileInfo.lintSpanMap);

    }

    private boolean isTopLevelDecl(JCTree tree) {
        return tree.getTag() == Tag.MODULEDEF
            || tree.getTag() == Tag.PACKAGEDEF
            || tree.getTag() == Tag.CLASSDEF;
    }

// FileInfo

    /**
     * Holds the calculated {@link Lint}s for top-level declarations in some source file.
     *
     * <p>
     * Instances evolve through these states:
     * <ul>
     *  <li>Before the file has been completely parsed, {@code #parsed} is false and {@link #lintSpanMap} is empty.
     *  <li>Immediately after the file has been parsed, {@code #parsed} is true and {@link #lintSpanMap} contains
     *      zero or more entries corresponding to the top-level declarations in the file, but whose values are null.
     *  <li>As each top-level declaration is attributed, the entries in {@link #lintSpanMap} are updated to non-null.
     * </ul>
     */
    private static class FileInfo {

        final Map<Span, List<LintSpan>> lintSpanMap = new HashMap<>();
        boolean parsed;

        // Find the top-level declaration containing the given position
        Span findDeclSpan(DiagnosticPosition pos) {
            return lintSpanMap.keySet().stream()
              .filter(span -> span.contains(pos))
              .findFirst()
              .orElse(null);
        }

        // Find the narrowest span in the given list that contains the given position
        static Optional<LintSpan> bestMatch(List<LintSpan> lintSpans, DiagnosticPosition pos) {
            int position = pos.getStartPosition();
            Assert.check(position != Position.NOPOS);
            LintSpan bestSpan = null;
            for (LintSpan lintSpan : lintSpans) {
                if (lintSpan.contains(position) && (bestSpan == null || bestSpan.contains(lintSpan))) {
                    bestSpan = lintSpan;
                }
            }
            return Optional.ofNullable(bestSpan);
        }
    }

// Span

    /**
     * Represents a lexical range in a file.
     *
     * @param tag tree node type (for debug only)
     * @param startPos starting position (inclusive)
     * @param endPos ending position (exclusive)
     * @param lint the applicable {@link Lint} configuration
     */
    private static class Span {

        final int startPos;
        final int endPos;

        Span(int startPos, int endPos) {
            this.startPos = startPos;
            this.endPos = endPos;
        }

        Span(JCTree tree) {
            this(TreeInfo.getStartPos(tree), TreeInfo.endPos(tree));
        }

        boolean contains(int pos) {
            return pos == startPos || (pos > startPos && pos < endPos);
        }

        boolean contains(DiagnosticPosition pos) {
            return contains(pos.getStartPosition());
        }

        boolean contains(Span that) {
            return this.startPos <= that.startPos && this.endPos >= that.endPos;
        }

        @Override
        public int hashCode() {
            return Integer.hashCode(startPos) ^ Integer.hashCode(endPos);
        }

        @Override
        public boolean equals(Object obj) {
            if (obj == this)
                return true;
            if (obj == null || obj.getClass() != this.getClass())
                return false;
            final Span that = (Span)obj;
            return this.startPos == that.startPos && this.endPos == that.endPos;
        }

        @Override
        public String toString() {
            return String.format("Span[%d-%d]", startPos, endPos);
        }
    }

// LintSpan

    /**
     * Represents a lexical range and the {@link Lint} configuration that applies therein.
     */
    private static class LintSpan extends Span {

        final Lint lint;

        LintSpan(int startPos, int endPos, Lint lint) {
            super(startPos, endPos);
            this.lint = lint;
        }

        LintSpan(JCTree tree, Lint lint) {
            super(tree);
            this.lint = lint;
        }

        // Note: no need for equals() or hashCode() here

        @Override
        public String toString() {
            return String.format("LintSpan[%d-%d, lint=%s]", startPos, endPos, lint);
        }
    }

// LintCalculator

    private static class LintCalculator extends TreeScanner {

        private final List<LintSpan> lintSpans = new ArrayList<>();
        private Lint currentLint;

        List<LintSpan> calculate(JCTree tree, Lint lint) {
            currentLint = lint;
            scan(tree);
            return lintSpans;
        }

        @Override
        public void visitModuleDef(JCModuleDecl tree) {
            scanDecl(tree, tree.sym, super::visitModuleDef);
        }

        @Override
        public void visitPackageDef(JCPackageDecl tree) {
            scanDecl(tree, tree.packge, super::visitPackageDef);
        }

        @Override
        public void visitClassDef(JCClassDecl tree) {
            scanDecl(tree, tree.sym, super::visitClassDef);
        }

        @Override
        public void visitMethodDef(JCMethodDecl tree) {
            scanDecl(tree, tree.sym, super::visitMethodDef);
        }

        @Override
        public void visitVarDef(JCVariableDecl tree) {
            scanDecl(tree, tree.sym, super::visitVarDef);
        }

        private <T extends JCTree> void scanDecl(T tree, Symbol symbol, Consumer<? super T> recursion) {
            Lint previousLint = currentLint;
            currentLint = Optional.ofNullable(symbol)   // symbol can be null if there were earlier errors
              .map(currentLint::augment)
              .orElse(currentLint);
            recursion.accept(tree);
            if (currentLint != previousLint) {          // Lint.augment() returns the same object if no change
                lintSpans.add(new LintSpan(tree, currentLint));
                currentLint = previousLint;
            }
        }
    }
}
