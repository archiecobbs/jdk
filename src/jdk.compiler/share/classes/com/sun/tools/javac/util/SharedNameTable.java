/*
 * Copyright (c) 1999, 2021, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package com.sun.tools.javac.util;

import java.lang.ref.SoftReference;

import com.sun.tools.javac.util.DefinedBy.Api;

/**
 * Implementation of Name.Table that stores all names in a single shared
 * byte array, expanding it as needed. This avoids the overhead incurred
 * by using an array of bytes for each name.
 *
 *  <p><b>This is NOT part of any supported API.
 *  If you write code that depends on this, you do so at your own risk.
 *  This code and its internal interfaces are subject to change or
 *  deletion without notice.</b>
 */
public class SharedNameTable extends Name.Table {
    // maintain a freelist of recently used name tables for reuse.
    private static List<SoftReference<SharedNameTable>> freelist = List.nil();

    public static synchronized SharedNameTable create(Names names) {
        while (freelist.nonEmpty()) {
            SharedNameTable t = freelist.head.get();
            freelist = freelist.tail;
            if (t != null) {
                return t;
            }
        }
        return new SharedNameTable(names);
    }

    private static synchronized void dispose(SharedNameTable t) {
        freelist = freelist.prepend(new SoftReference<>(t));
    }

    /** The hash table for names.
     */
    private NameImpl[] hashes;

    /** The shared byte array holding all encountered names.
     */
    public byte[] bytes;

    /** The mask to be used for hashing
     */
    private int hashMask;

    /** The number of filled bytes in `names'.
     */
    private int nc = 0;

    /** Allocator
     *  @param names The main name table
     *  @param hashSize the (constant) size to be used for the hash table
     *                  needs to be a power of two.
     *  @param nameSize the initial size of the name table.
     */
    public SharedNameTable(Names names, int hashSize, int nameSize) {
        super(names);
        hashMask = hashSize - 1;
        hashes = new NameImpl[hashSize];
        bytes = new byte[nameSize];

    }

    public SharedNameTable(Names names) {
        this(names, 0x8000, 0x20000);
    }

    @Override
    public Name fromChars(char[] cs, int start, int len) {
        int nc = this.nc;
        byte[] bytes = this.bytes = ArrayUtils.ensureCapacity(this.bytes, nc + len * 3);
        int nbytes = Convert.chars2utf(cs, start, bytes, nc, len) - nc;
        int h = hashValue(bytes, nc, nbytes) & hashMask;
        NameImpl n = hashes[h];
        while (n != null &&
                (n.getByteLength() != nbytes ||
                !equals(bytes, n.index, bytes, nc, nbytes))) {
            n = n.next;
        }
        if (n == null) {
            n = newNameImpl(nc, nbytes);
            n.next = hashes[h];
            hashes[h] = n;
            this.nc = nc + nbytes;
            if (nbytes == 0) {
                this.nc++;
            }
        }
        return n;
    }

    @Override
    public Name fromUtf(byte[] cs, int start, int len) {
        int h = hashValue(cs, start, len) & hashMask;
        NameImpl n = hashes[h];
        byte[] names = this.bytes;
        while (n != null &&
                (n.getByteLength() != len || !equals(names, n.index, cs, start, len))) {
            n = n.next;
        }
        if (n == null) {
            int nc = this.nc;
            names = this.bytes = ArrayUtils.ensureCapacity(names, nc + len);
            System.arraycopy(cs, start, names, nc, len);
            n = newNameImpl(nc, len);
            n.next = hashes[h];
            hashes[h] = n;
            this.nc = nc + len;
            if (len == 0) {
                this.nc++;
            }
        }
        return n;
    }

    @Override
    public void dispose() {
        dispose(this);
    }

    // Build and return an AsciiNameImpl, if possible, otherwise a regular NameImpl
    NameImpl newNameImpl(final int index, final int length) {
        int offset = index;
        int remain = length;
        while (remain > 0) {
            if ((bytes[offset++] & 0x80) != 0)
                break;
            remain--;
        }
        return remain == 0 ?
          new AsciiNameImpl(this, index, length) :
          new NameImpl(this, index, length);
    }

    // Implementation used when there are one or more non-ASCII characters
    static class NameImpl extends Name {
        /** The next name occupying the same hash bucket.
         */
        NameImpl next;

        /** The index where the bytes of this name are stored in the global name
         *  buffer `byte'.
         */
        int index;

        /** The number of bytes in this name.
         */
        int length;

        NameImpl(SharedNameTable table, int index, int length) {
            super(table);
            this.index = index;
            this.length = length;
        }

        @Override
        public int getIndex() {
            return index;
        }

        @Override
        public int getByteLength() {
            return length;
        }

        @Override
        public byte getByteAt(int i) {
            return getByteArray()[index + i];
        }

        @Override
        public byte[] getByteArray() {
            return ((SharedNameTable) table).bytes;
        }

        @Override
        public int getByteOffset() {
            return index;
        }

        /** Return the hash value of this name.
         */
        @DefinedBy(Api.LANGUAGE_MODEL)
        public int hashCode() {
            return index;
        }

        /** Is this name equal to other?
         */
        @DefinedBy(Api.LANGUAGE_MODEL)
        public boolean equals(Object other) {
            return (other instanceof Name name)
                    && table == name.table
                    && index == name.getIndex();
        }
    }

    // Implementation used when there are only ASCII (0x00 - 0x7f) characters
    static class AsciiNameImpl extends NameImpl {

        AsciiNameImpl(SharedNameTable table, int index, int length) {
            super(table, index, length);
        }

        @Override
        public boolean contentEquals(CharSequence cs) {
            if (cs.length() != length)
                return false;
            byte[] data = getByteArray();
            int off = index;
            for (int i = 0; i < length; i++) {
                char ch = cs.charAt(i);
                if ((int)ch != (data[off++] & 0xff))
                    return false;
            }
            return true;
        }

        @Override
        public char charAt(int pos) {
            if (pos < 0 || pos >= length)
                throw new IndexOutOfBoundsException();
            return (char)(getByteArray()[index + pos] & 0xff);
        }

        @Override
        public int length() {
            return length;
        }

        @Override
        public CharSequence subSequence(int start, int end) {
            if (start < 0 || end < start || end > length)
                throw new IndexOutOfBoundsException();
            return new AsciiCharSequence(getByteArray(), index + start, end - start);
        }

        @Override
        public String toString() {
            return AsciiNameImpl.toString(getByteArray(), index, length);
        }

        static String toString(byte[] data, int offset, int length) {
            char[] array = new char[length];
            for (int i = 0; i < length; i++)
                array[i] = (char)(data[offset + i] & 0xff);
            return new String(array);
        }
    }

// AsciiCharSequence

    static class AsciiCharSequence implements CharSequence {

        private final byte[] data;
        private final int offset;
        private final int length;

        AsciiCharSequence(byte[] data, int offset, int length) {
            this.data = data;
            this.offset = offset;
            this.length = length;
        }

        @Override
        public char charAt(int index) {
            if (index < 0 || index >= length)
                throw new IndexOutOfBoundsException();
            return (char)(data[offset + index] & 0xff);
        }

        @Override
        public int length() {
            return length;
        }

        @Override
        public CharSequence subSequence(int start, int end) {
            if (start < 0 || end < start || end > length)
                throw new IndexOutOfBoundsException();
            return new AsciiCharSequence(data, offset + start, end - start);
        }

        @Override
        public String toString() {
            return AsciiNameImpl.toString(data, offset, length);
        }
    }
}
