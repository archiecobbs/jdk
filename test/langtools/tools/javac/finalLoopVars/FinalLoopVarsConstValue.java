/*
 * @test /nodynamiccopyright/
 * @bug 8338711
 * @summary Test final for() loop variable support doesn't permit illegal stuff
 * @compile/fail/ref=FinalLoopVarsConstValue.out -XDrawDiagnostics FinalLoopVarsConstValue.java
 * @enablePreview
 */

import java.util.function.*;

public class FinalLoopVarsConstValue {

    // Test: a final for() loop variable that's not mutated has a constant value
    {
        for (final int i = 0; i != 0; /* no change */)
            System.out.println("impossible");   // error: unreachable statement
        System.out.println("ok");
    }

    // Test: a final for() loop variable that's not mutated has a constant value
    {
        for (final int i = 0; i == 0; /* no change */)
            System.out.println("ok");
        System.out.println("impossible");       // error: unreachable statement
    }

    // Test: weird nesting doesn't confuse the analysis logic
    {
        for (final int i = 0, j = new IntSupplier() {
                                    public int getAsInt() {
                                        for (int i = 0; i < 3; i++)     // mutate a different "i" here
                                            System.out.println(i++);
                                        return 0;
                                    }
                                  }.getAsInt();
          i != 0;
          /* no change */)
            System.out.println("impossible");   // error: unreachable statement
    }
}
