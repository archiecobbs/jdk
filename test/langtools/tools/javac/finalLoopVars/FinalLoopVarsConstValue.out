FinalLoopVarsConstValue.java:16:13: compiler.err.unreachable.stmt
FinalLoopVarsConstValue.java:24:9: compiler.err.unreachable.stmt
FinalLoopVarsConstValue.java:38:13: compiler.err.unreachable.stmt
3 errors
